core/server/common/src/main/java/alluxio/master/journal/ufs/UfsJournal.java

```diff
@@ -14,133 +14,384 @@
+  }
+
+  @Override
+  public JournalWriter getWriter(JournalWriterOptions options) throws IOException {
+    if (options.isPrimary()) {
+      return new UfsJournalLogWriter(this, options);
+    } else {
+      return new UfsJournalCheckpointWriter(this, options);
+    }
+  }
+
+  @Override
+  public boolean isFormatted() throws IOException {
+    UnderFileSystem ufs = UnderFileSystem.Factory.get(mLocation.toString());
+    UnderFileStatus[] files = ufs.listStatus(mLocation.toString());
+    if (files == null) {
+      return false;
     }
+    // Search for the format file.
+    String formatFilePrefix = Configuration.get(PropertyKey.MASTER_FORMAT_FILE_PREFIX);
+    for (UnderFileStatus file : files) {
+      if (file.getName().startsWith(formatFilePrefix)) {
+        return true;
+      }
+    }
+    return false;
+  }
+
+  @Override
+  public void format() throws IOException {
+    URI location = getLocation();
+    LOG.info("Formatting {}", location);
+    if (mUfs.isDirectory(location.toString())) {
+      for (UnderFileStatus p : mUfs.listStatus(location.toString())) {
+        URI childPath = URIUtils.appendPathOrDie(location, p.getName());
+        boolean failedToDelete;
+        if (p.isDirectory()) {
+          failedToDelete = !mUfs
+              .deleteDirectory(childPath.toString(), DeleteOptions.defaults().setRecursive(true));
+        } else {
+          failedToDelete = !mUfs.deleteFile(childPath.toString());
+        }
+        if (failedToDelete) {
+          throw new IOException(String.format("Failed to delete %s", childPath));
+        }
+      }
+    } else if (!mUfs.mkdirs(location.toString())) {
+      throw new IOException(String.format("Failed to create %s", location));
+    }
+
+    // Create a breadcrumb that indicates that the journal folder has been formatted.
+    UnderFileSystemUtils.touch(URIUtils.appendPathOrDie(location,
+        Configuration.get(PropertyKey.MASTER_FORMAT_FILE_PREFIX) + System.currentTimeMillis())
+        .toString());
   }
 
   /**
-   * @return the location of the journal checkpoint
+   * @return the log directory location
    */
-  protected URI getCheckpoint() {
-    try {
-      return URIUtils.appendPath(mLocation, CHECKPOINT_FILENAME);
-    } catch (URISyntaxException e) {
-      throw new RuntimeException(e);
-    }
+  URI getLogDir() {
+    return mLogDir;
   }
 
   /**
-   * @return the location of the current log
+   * @return the checkpoint directory location
    */
-  public URI getCurrentLog() {
-    try {
-      return URIUtils.appendPath(mLocation, ENTRY_LOG_FILENAME_BASE + CURRENT_LOG_EXTENSION);
-    } catch (URISyntaxException e) {
-      throw new RuntimeException(e);
-    }
+  URI getCheckpointDir() {
+    return mCheckpointDir;
+  }
+
+  /**
+   * @return the temporary directory location
+   */
+  URI getTmpDir() {
+    return mTmpDir;
+  }
+
+  /**
+   * @return the under file system instance
+   */
+  UnderFileSystem getUfs() {
+    return mUfs;
+  }
```

Comment:
an alternative is to pass the UFS as an argument into all the methods that currently extract it from UfsJournal